[community tutorial] (ipynb) time series visualization with altair

jdbcode opened the pull request.

gino-m commented:
Hi! Is this tutorial still WIP? Note that integration tests were failing due to the following:
```
Failures:
1) tutorials / time-series-visualization-with-altair / tutorial.ipynb relative src paths in html
  Message:
    AssertionError [ERR_ASSERTION]: absolute src paths not allowed: src=\"https:/
  Stack:
    error properties: Object({ generatedMessage: false, code: 'ERR_ASSERTION', actual: false, expected: true, operator: '==' })
        at <Jasmine>
        at UserContext.<anonymous> (/home/travis/build/google/earthengine-community/tutorials/test/tutorials.test.js:170:11)
        at <Jasmine>
        at runCallback (timers.js:705:18)
        at tryOnImmediate (timers.js:676:5)
        at processImmediate (timers.js:658:5)
2) tutorials / time-series-visualization-with-altair / tutorial.ipynb copyright header
  Message:
    AssertionError [ERR_ASSERTION]: first cell must contain valid copyright header
  Stack:
    error properties: Object({ generatedMessage: false, code: 'ERR_ASSERTION', actual: false, expected: true, operator: '==' })
        at <Jasmine>
        at UserContext.<anonymous> (/home/travis/build/google/earthengine-community/tutorials/test/tutorials.test.js:146:11)
        at <Jasmine>
        at runCallback (timers.js:705:18)
        at tryOnImmediate (timers.js:676:5)
        at processImmediate (timers.js:658:5)
```

jdbcode commented:
> Hi! Is this tutorial still WIP? Note that integration tests were failing due to the following:

Yes, still WIP. I need to address your comments and fix the integration test issues, then have @tylere take a look. Planning on spending some time on it in the next 2 weeks.

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Good point. I've changed the sentence a little to avoid using "scope of the function". The need for the global dict is to be able to call on it from a function that is being mapped over and image collection, which can only accept a single param - the current image. In later steps we mutate the dict as needed for each application of the mapped <code>reduceRegion</code>  function.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>It was the fasted way to convert from JS to Py. I've corrected in in all cases.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>No, it is a mix of image sources that changes by zoom and location. Upon further investigation it also appears that use of these tiles in this application might violate terms of service. I've opted for using a USDA NAIP image, which is a more straight forward attribution.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
@gino-m Thanks so much for the review! I've addressed all comments - two are marked unresolved - thought you might want to take a look at the changes. Note that there was a lot of poor Python code formatting, I've used the [Google style guide](https://google.github.io/styleguide/pyguide.html) this time around. Will deal with integration test issues at the end of the review (@tylere PTAL for consistency with EE-Py best practices).

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>Do we want to links to authors? i.e.</p><ul><li>Author: <a href="https://github.com/jdbcode" target="_blank">jdbcode</a></li></ul><p> </p><p>Suggestion to make use of articles consistent:</p><ul><li>Formatting a table in Earth Engine</li><li>Transferring <u>an</u> Earth Engine table to <u>a</u> Colab Python kernel</li><li>Converting <u>an</u> Earth Engine table to&nbsp;<u>a</u> <a href="https://pandas.pydata.org/" target="_blank">pandas</a>&nbsp;DataFrame</li></ul><p> </p><p>The following seems unnecessary, given that you are already linking to the page in the previous sentence.</p><p>"For a review of Python syntax as it relates to the Earth Engine API, please see this <a href="https://developers.google.com/earth-engine/python_install#syntax" target="_blank">guide</a>."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

jdbcode replied:
<p>Re: articles - Done here and throughout.</p><p> </p><p>Re: remove redundant syntax link - Done.</p>

tylere replied:
<p>"Earth an Engine table" -&gt; "an Earth Engine table"</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>Suggested change:</p><p><strong>Earth Engine</strong> is used to access, clean, and reduce large amounts of spatiotemporal data, <strong>pandas</strong> is used to analyze and organize the results, and</p><p> </p><p>Suggested change:</p><p><strong>Note:</strong> This notebook demonstrates an analysis template and interactive workflow that is appropriate for a certain size of dataset, but there are limitations to interactive computation time and server-to-client data transfer size imposed by Colab and Earth Engine.  To analyze even larger datasets, you may need to modify the workflow to <a href="https://developers.google.com/earth-engine/exporting#exporting-tables-and-vector-data" target="_blank">export</a> <code>featureCollection</code> results from Earth Engine as static assets and then use the static assets to perform the subsequent steps involving Earth Engine table formatting, conversion to pandas DataFrame, and charting with Altair.</p><p> </p><p>It would be better if the export link pointed to a Python export example:</p><p><a href="https://developers.google.com/earth-engine/python_install#exporting-data" target="_blank">https://developers.google.com/earth-engine/python_install#exporting-data</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

jdbcode replied:
<p>Done.</p><p> </p><p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>The spacing between bullets is inconsistent. Strange...</p><p> </p><p>The PDSI link should point to <a href="https://developers.google.com/earth-engine/datasets/catalog/IDAHO_EPSCOR_PDSI" target="_blank">https://developers.google.com/earth-engine/datasets/catalog/IDAHO_EPSCOR_PDSI</a></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>Should "Python" be changed to "pandas" or "Altair"? (Python is used to control Earth Engine in Steps 1-3.)</p><p> </p><p>Should "first example" be "first dataset"?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

jdbcode replied:
<p>Re: steps - I wanted to give folks a sense for what was happening in Earth Engine and Python. What I really meant was server-side and client-side. I've updated to use "server-side Earth Engine" and "client-side &lt;py library&gt;"</p><p> </p><p>Re: "first example" &gt; "first dataset" - done.</p><p> </p>

tylere replied:
<p>lgtm</p><p> </p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>Suggest providing more details on the libraries:</p><ul><li><a href="https://pandas.pydata.org/" target="_blank"><strong>pandas</strong></a>: data analysis (including the <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html" target="_blank">DataFrame</a> data structure)</li><li><a href="https://altair-viz.github.io/" target="_blank"><strong>altair</strong></a>: declarative visualization library (used for charting)</li><li><a href="https://numpy.org/" target="_blank"><strong>numpy</strong></a>: array-processing package (used for linear regression)</li><li><a href="https://python-visualization.github.io/folium/" target="_blank"><strong>folium</strong></a>: interactive web map</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>I think that setting global default args that are overwritten as you proceed through the notebook may be confusing to readers and error-prone, and is not an approach that we want to highlight. I think a better approach would be to create a function that returns functions that can be mapped over image collections. For example:</p><p> </p><p> </p><p><code>def create_reduce_region_function(geometry,</code></p><p><code>                                  reducer=ee.Reducer.mean(),</code></p><p><code>                                  crs='EPSG:4326',</code></p><p><code>                                  scale=1000,</code></p><p><code>                                  bestEffort=True,</code></p><p><code>                                  maxPixels=1e14,</code></p><p><code>                                  tileScale=4</code></p><p><code>                                  ): </code></p><p><code>    def reduce_region_function(img):</code></p><p><code>      stat = img.reduceRegion(</code></p><p><code>        reducer=reducer,</code></p><p><code>        geometry=geometry,</code></p><p><code>        scale=scale,</code></p><p><code>        crs=crs,</code></p><p><code>        bestEffort=bestEffort,</code></p><p><code>        maxPixels=maxPixels,</code></p><p><code>        tileScale=tileScale)</code></p><p><code>      return ee.Feature(geometry, stat).set({'millis': img.date().millis()})</code></p><p> </p><p><code>    return reduce_region_function</code></p><p> </p><p> </p><p><code>pdsi_function = create_reduce_region_function(scale=5000,</code></p><p><code>                                              reducer=ee.Reducer.mean(),</code></p><p><code>                                              geometry=aoi,</code></p><p><code>                                              crs='EPSG:3310')</code></p><p> </p><p> </p><p><code>pdsi_stat_fc = ee.FeatureCollection(pdsi.map(pdsi_function)).filter(</code></p><p><code>    ee.Filter.notNull(pdsi.first().bandNames()))</code></p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

jdbcode replied:
<p>This is much better - thanks for the suggestion! It has been implemented - done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>"map this over" -&gt; "map this function over"</p><p> </p><p>I wonder of the details of 1 and 2 would be better presented in the docstring of the function. Writing function <a href="https://google.github.io/styleguide/pyguide.html#381-docstrings" target="_blank">docstrings</a> is a best practice I think we should encourage.</p><p> </p><p>Suggest explaining why <code>.set({'millis': img.date().millis()</code>is needed.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>Suggest referring back to Figure 2a, which I assume shows this aoi geometry.  You might also want to include an interactive folium map showing the aoi, since you use folium later in the tutorial.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

jdbcode replied:
<p>Done - opted for figure reference.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>"Print it to see how it is formatted." -&gt; "Print <u>a small part</u> to see how it is formatted."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>"to pandas DataFrame" -&gt; "to a pandas DataFrame"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p><code>display(pdsi_df)</code></p><p><code>print('\n\n', pdsi_df.dtypes)</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>Suggest using an interactive chart, that highlights the selected year.</p><p> </p><p> </p><p><code># Chart NDVI time series as an Altair line chart.</code></p><p> </p><p><code>highlight = alt.selection(type='single', on='mouseover',</code></p><p><code>                          fields=['Year'], nearest=True)</code></p><p> </p><p><code>base = alt.Chart(ndvi_df).encode(</code></p><p><code>  alt.X('DOY:O'),</code></p><p><code>  alt.Y('NDVI:Q', scale=alt.Scale(domain=(0.1, 0.7))),</code></p><p><code>  alt.Color('Year:O', scale=alt.Scale(scheme='magma'))</code></p><p><code>)</code></p><p> </p><p><code>points = base.mark_circle().encode(</code></p><p><code>    opacity=alt.value(0),</code></p><p><code>    tooltip=[</code></p><p><code>        alt.Tooltip('Year:O', title='Year'),</code></p><p><code>        alt.Tooltip('DOY:O', title='DOY'),</code></p><p><code>        alt.Tooltip('NDVI:Q', title='NDVI')</code></p><p><code>    ]</code></p><p><code>).add_selection(</code></p><p><code>    highlight</code></p><p><code>)</code></p><p> </p><p><code>lines = base.mark_line().encode(</code></p><p><code>    size=alt.condition(~highlight, alt.value(1), alt.value(3))</code></p><p><code>)</code></p><p> </p><p><code>(points + lines).properties(width=600, height=300).interactive()</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='72'/>

jdbcode replied:
<p>Great idea! Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p><s>, though it's a bit of a spaghetti mess to interpret.</s></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='73'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>Alternative chart configuration using a "base" chart, that the points and fit charts inherit from.</p><p>(I think this is a bit easier to read and understand, but I don't feel very strongly about it.)</p><p> </p><p> </p><p><code>base = alt.Chart(ndvi_pdsi_df).encode(</code></p><p><code>    x=alt.X('PDSI:Q', scale=x_scale)</code></p><p><code>)</code></p><p> </p><p><code>points = base.mark_circle(size=60).encode(</code></p><p><code>    y=alt.Y('NDVI:Q', scale=y_scale),</code></p><p><code>    color=alt.Color('Year:O', scale=alt.Scale(scheme='magma')),</code></p><p><code>    tooltip=['Year', 'PDSI', 'NDVI'])</code></p><p> </p><p><code>fit = base.mark_line().encode(</code></p><p><code>    y=alt.Y('Fit:Q', scale=y_scale),</code></p><p><code>    color=alt.value('#808080'))</code></p><p> </p><p><code>(fit + points).properties(width=600, height=300).interactive()</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='89'/>

jdbcode replied:
<p>This is great, I've reused the "base" configuration throughout. Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>Bonus points if you figure out how to calculate and plot the <a href="https://stackoverflow.com/questions/27164114/show-confidence-limits-and-prediction-limits-in-scatter-plot" target="_blank">95% confidence limits</a> for this regression. Certainly not necessary though...</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='90'/>

jdbcode replied:
<p>I think it is beyond the scope of this tutorial, but it would be neat to do a tutorial that focuses on sampling, time series, and stats.</p><p> </p><p>This snippet might work:</p><p> </p><pre class="ql-syntax" spellcheck="false"># example data
x = np.array([4.0,2.5,3.2,5.8,7.4,4.4,8.3,8.5])
y = np.array([2.1,4.0,1.5,6.3,5.0,5.8,8.1,7.1])
&nbsp;
# fit a curve to the data using a least squares 1st order polynomial fit
z = np.polyfit(x,y,1)
p = np.poly1d(z)
fit = p(x)
&nbsp;
# get the coordinates for the fit curve
c_y = [np.min(fit),np.max(fit)]
c_x = [np.min(x),np.max(x)]
&nbsp;
# predict y values of origional data using the fit
p_y = z[0] * x + z[1]
&nbsp;
# calculate the y-error (residuals)
y_err = y -p_y
&nbsp;
# create series of new test x-values to predict for
p_x = np.arange(np.min(x),np.max(x)+1,1)
&nbsp;
# now calculate confidence intervals for new test x-series
mean_x = np.mean(x)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # mean of x
n = len(x)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # number of samples in origional fit
t = 2.31&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; # appropriate t value (where n=9, two tailed 95%)
s_err = np.sum(np.power(y_err,2))&nbsp;&nbsp; # sum of the squares of the residuals
&nbsp;
confs = t * np.sqrt((s_err/(n-2))*(1.0/n + (np.power((p_x-mean_x),2)/
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;((np.sum(np.power(x,2)))-n*(np.power(mean_x,2))))))
&nbsp;
# now predict y based on test x-values
p_y = z[0]*p_x+z[0]
&nbsp;
# get lower and upper confidence limits based on predicted y and confidence intervals
lower = p_y - abs(confs)
upper = p_y + abs(confs)
</pre><p> </p><p> </p>

tylere replied:
Nice! I agree it is beyond the scope of this tutorial, and a sampling/stats tutorial in the future may be worth writing.

I moved the your snipped to a new notebook and added a chart:
https://colab.sandbox.google.com/drive/1ip37YBzAdq0Dq0RfIKM9ZggAvkrN2pm4

jdbcode replied:
Thanks for moving it to a NB and for adding the chart!

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>Suggest adding the Sierra Nevada ecoregion as a reference layer, to connect this part of the tutorial with the prior content.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='93'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>Suggest expanding this text block to define and link to terms needed to understand the following code block, which contains acronyms that may not be familiar to readers.</p><ul><li>OLI</li><li>TM</li><li>ETM+</li><li>CFMask</li><li>NBR</li><li>understanding Earth Engine joins</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='98'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p> </p><p> </p><p><code>nbr_function = create_reduce_region_function(scale=5000,</code></p><p><code>                                              reducer=ee.Reducer.first(),</code></p><p><code>                                              geometry=point,</code></p><p><code>                                              crs='EPSG:3310')</code></p><p> </p><p><code>nbr_stat_fc = ee.FeatureCollection(landsat_col.map(nbr_function)).filter(</code></p><p><code>    ee.Filter.notNull(landsat_col.first().bandNames()))</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='102'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>This ends up being strangely formatted in Colab, with the  "Note" starting to the left of the numbers.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='103'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p><code>display(nbr_df)</code></p><p><code>print('\n\n', nbr_df.dtypes)</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='104'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p><code>dcp_function = create_reduce_region_function(scale=5000,</code></p><p><code>reducer=ee.Reducer.first(),</code></p><p><code>geometry=point,</code></p><p><code>crs='EPSG:3310')</code></p><p> </p><p><code>dcp_stat_fc = ee.FeatureCollection(dcp_col.map(dcp_function)).filter(</code></p><p><code>ee.Filter.notNull(dcp_col.first().bandNames()))</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='115'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p><code>display(dcp_df)</code></p><p><code>print('\n\n', dcp_df.dtypes)</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='117'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p><code>reduce_region</code> -&gt; <code>dcp_function</code></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='124'/>

jdbcode replied:
<p>Done.</p>

tylere commented on tutorials/time-series-visualization-with-altair/tutorial.ipynb:
```diff
@@ -0,0 +1,2133 @@
+{
```
<p>Suggest keeping the encoding parameters on separate lines, to improve readability, similar to earlier code blocks.</p><p> </p><p><code># Define a line plot by reducing intra-annual observations to median.</code></p><p><code>line = alt.Chart(climate_df).mark_line().encode(</code></p><p><code>    x='Year:O',</code></p><p><code>    y=alt.Y('median(Precip-mm):Q', title='Precipitation (mm/month)'),</code></p><p><code>    color='Model')</code></p><p> </p><p><code># Define an interquartile range plot to describe the intra-annual distribution.</code></p><p><code>band = alt.Chart(climate_df).mark_errorband(extent='iqr').encode(</code></p><p><code>    x='Year:O',</code></p><p><code>    y=alt.Y('Precip-mm:Q', title='Precipitation (mm/month)'),</code></p><p><code>    color='Model')</code></p><p> </p><p><code># Combine the band and line plots; display it.</code></p><p><code>(band + line).properties(width=600, height=300)</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='134'/>

jdbcode replied:
<p>Done.</p>

jdbcode commented:
<p>Tyler suggested using a function that returns a reduction function to avoid the global param dictionary. I've updated this section, the text is no longer included.</p>

---

 <a href='https://app.reviewnb.com/google/earthengine-community/pull/152/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

jdbcode commented:
@tylere and @gino-m thank you for the reviews - this tutorial is in much better shape! All of your comments have been addressed. Can you please take another look and mark approved or recommend additional changes. Note that I might still need to address integration test issues, but would like to do that once the tutorial content is settled.

tylere commented:
I only noticed one remaining typo:
"Converting **Earth an Engine** table" -> "Converting an Earth Engine table"

LGTM

jdbcode commented:
Hi @tylere, I've fixed the typo and lint errors related to use of absolute <img> source paths. Can you please approve the PR.
